load_data.py

```diff
@@ -0,0 +1,159 @@
+from datasets import (
+    DatasetDict,
+    get_dataset_infos as get_ds_infos,
+    get_dataset_split_names as get_ds_split_names,
+    load_dataset as load_ds)
+from os import environ
+from tokenizers import decoders, pre_tokenizers, processors, Tokenizer
+from tokenizers.models import BPE
+from tokenizers.trainers import BpeTrainer
+from torch.utils.data import DataLoader
+from transformers import PreTrainedTokenizerFast
+
+# Disable parallelism to avoid errors with DataLoaders later on
+environ["TOKENIZERS_PARALLELISM"] = "false"
+
+def get_loaders_tokenizer(
+        dataset_name: str,
+        seq_len: int,
+        batch_size: int,
+        vocab_size: int,
+        data_dir: str,
+        dataset_config: str=None,
+        text_feature: str="text",
+        max_token_len: int=20,
+        splits: list[float]=[0.7, 0.2, 0.1],
+        rand_seed: int=None) -> \
+            tuple[DataLoader, DataLoader, DataLoader, Tokenizer]:
+    """ Loads Hugging Face dataset and creates DataLoaders and Tokenizer.
+    Args:
+        dataset_name (str): Name of Hugging Face dataset.
+        seq_len (int): Context window/sequence length.
+        batch_size (int): Batch size.
+        vocab_size (int): Maximum vocabulary size.
+        data_dir (str): Relative path from base of repository to directory in
+            which to download the dataset.
+        dataset_config (str): Configuration/subset of dataset to use.
+        text_feature (str): Name of the feature/column of the dataset to use.
+        max_token_len (int): Prevents tokenizer creating tokens longer than the
+            specified size.
+        splits (list[float]): A list of three floats containing the train,
+            validation, and test splits respectively. Should sum to 1.
+        rand_seed (int): Seed used during dataset shuffling, ignored if None.
+
+    Returns:
+        Tuple with the format: (Training DataLoader, Validation DataLoader,
+        Testing DataLoader, Tokenizer object).
+    """
+    # Test text_feature is actually a feature of the dataset
+    ds_features = get_ds_infos(
+        dataset_name,
+        trust_remote_code=True)[dataset_config].features
+    assert text_feature in ds_features, \
+        f"'{text_feature}' not in '{dataset_name}' features {ds_features}!"
+
+    # Retrieve iterators for each split of the dataset
+    entire_dataset = load_ds(
+        path=dataset_name,
+        name=dataset_config,
+        split="all",
+        cache_dir=data_dir,
+        trust_remote_code=True)
+
+    # Function to filter out undesired inputs. In this case, filter out
+    # instances with only whitespace
+    filter_fun = lambda inst_dict : bool(inst_dict[text_feature].strip())
+
+    # Filter out undesired data instances
+    entire_dataset = entire_dataset.filter(filter_fun)
+
+    # Split into training, validation, and testing datasets
+    train_testvalid = entire_dataset.train_test_split(
```

Comment:
My understanding of this code is you split two times using the train_test_split() function.
train = train_testvalid["train"]
validation = test_valid["train"]
test = test_valid["test"]

However, implicit in the ordering of the variable name train_testvalid is that train is under ["train"], and testvalid is under ["test"].
My issue is that test_valid would implicitly suggest that test is under ["train"], and validation is under ["test"]. The code says otherwise.
**If you could rename the variables train_testvalid and test_valid to be train_validtest and valid_test, respectively, that would make the code a bit clearer to read.** Furthermore, it would follow the order the comment above states (#split into training, validation, and testing datasets).

Reply:
Good catch! Even reviewing this myself after you pointed it out made me confused for a minute. I like your idea of renaming, thereby attempting to keep things clearer. I've made this change in [af42f88](https://github.com/DRAGNLabs/301r_retnet/pull/15/commits/af42f88d521e8896ce9b9803b8b641a256664662).